lib/puppet/provider/service/smf.rb

```diff
@@ -160,11 +175,92 @@ def status
       raise Puppet::Error,
         "Unmanageable state '#{state}' on service #{self.name}"
     end
+  end
 
+  # Helper that encapsulates the clear + svcadm [enable|disable]
+  # logic in one place. Makes it easy to test things out and also
+  # cleans up flush's code.
+  def maybe_clear_service_then_svcadm(cur_state, subcmd, flags)
+    # If the cur_state is maint or degraded, then we need to clear the service
+    # before we enable or disable it.
+    adm('clear', self.service_fmri) if [:maintenance, :degraded].include?(cur_state)
+    adm(subcmd, flags, self.service_fmri)
   end
 
-  def stopcmd
-    [command(:adm), :disable, "-s", @resource[:name]]
+  # The flush method is necessary for the SMF provider because syncing the enable and ensure
+  # properties are not independent operations like they are in most of our other service
+  # providers.
+  def flush
```

Comment:
I hope this is clear. This came from hours of trying to simplify the code and I found out that this implementation seemed like it did a good job communicating what's going on. Please feel free to suggest an alternative approach.

Reply:
shouldn't self.instances take carte of @propery_hash for us?

Reply:
I don't understand what you're asking, can you clarify?

Reply:
From what I've been reading (http://garylarizza.com/blog/2013/12/15/seriously-what-is-this-provider-doing/, search for "A @property_hash is born…"), self.instances is where @propery_hash is managed. Just seeing if we are/should be using that model here, or if it doesn't apply.

Reply:
Actually, I think we should be fine here. `property_hash` is one way of managing the property values for flush to use -- see here https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/network_device.rb#L55 for example on how they update it. They retrieve the values for property hash here: https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/network_device.rb#L18 and https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/util/network_device/cisco/device.rb#L193. They use `prefetch`, which is what Puppet calls only if the provider class has it.

I *think* we can use `property_hash` too. It will just be empty for our case while we populate it with values. But I want to confirm with e.g. @branan if that is safe to do so. So some of what I originally thought about `@property_hash` was incorrect, but the model I'm using here is still correct. Just am not sure if it is necessary (i.e. if we can replace it with `property_hash` instead).

Reply:
👍